libqpdf/QPDF.cc

```diff
@@ -1293,62 +1293,28 @@ QPDF::showXRefTable()
 void
 QPDF::fixDanglingReferences(bool force)
 {
-    if (this->m->fixed_dangling_refs && (!force)) {
+    if (this->m->fixed_dangling_refs && !force) {
         return;
     }
-    this->m->fixed_dangling_refs = true;
-
-    // Create a set of all known indirect objects including those
-    // we've previously resolved and those that we have created.
-    std::set<QPDFObjGen> to_process;
-    for (auto const& iter: this->m->obj_cache) {
-        to_process.insert(iter.first);
-    }
-    for (auto const& iter: this->m->xref_table) {
-        to_process.insert(iter.first);
-    }
 
-    // For each non-scalar item to process, put it in the queue.
-    std::list<QPDFObjectHandle> queue;
-    queue.push_back(this->m->trailer);
-    for (auto const& og: to_process) {
-        auto obj = getObject(og);
-        if (obj.isDictionary() || obj.isArray()) {
-            queue.push_back(obj);
-        } else if (obj.isStream()) {
-            queue.push_back(obj.getDict());
-        }
-    }
-
-    // Process the queue by recursively resolving all object
-    // references. We don't need to do loop detection because we don't
-    // traverse known indirect objects when processing the queue.
-    while (!queue.empty()) {
-        QPDFObjectHandle obj = queue.front();
-        queue.pop_front();
-        std::list<QPDFObjectHandle> to_check;
-        if (obj.isDictionary()) {
-            std::map<std::string, QPDFObjectHandle> members =
-                obj.getDictAsMap();
-            for (auto const& iter: members) {
-                to_check.push_back(iter.second);
-            }
-        } else if (obj.isArray()) {
-            auto arr = QPDFObjectHandle::ObjAccessor::asArray(obj);
-            arr->addExplicitElementsToList(to_check);
-        }
-        for (auto sub: to_check) {
-            if (sub.isIndirect()) {
-                if ((sub.getOwningQPDF() == this) &&
-                    isUnresolved(sub.getObjGen())) {
-                    QTC::TC("qpdf", "QPDF detected dangling ref");
-                    queue.push_back(sub);
-                }
-            } else {
-                queue.push_back(sub);
+    if (!this->m->fixed_dangling_refs) {
+        for (auto const& iter: this->m->xref_table) {
+            auto og = iter.first;
+            if (!isCached(og)) {
+                m->obj_cache[og] =
+                    ObjCache(QPDF_Unresolved::create(this, og), -1, -1);
             }
         }
+        for (auto const& iter: this->m->obj_cache) {
+            resolve(iter.first);
+        }
+    }
+    // Final pass fixes dangling references reference.
```

Comment:
Here's where to put the explanation. Perhaps a comment in whatever part of the code is causing the dangling reference to appear in the object cache might be in order as well.

Reply:
except now I think this pass is not needed at all.